marketplaceItems/drawApp/draw_app.js

```diff
@@ -0,0 +1,504 @@
+    function onClicked() {
+        if (marker) {
+            button.editProperties({ isActive: false });
+            if (HMD.active) {
+                closeHMDMode();
+            } else {
+                closeDesktopMode();
+            }
+            Entities.deleteEntity(marker);
+            marker = null;
+            playSound(CLOSE_SOUND, CLOSE_SOUND_VOLUME, MyAvatar.position, true);
+        } else {
+            if (HMD.active) {
+                setUpHMDMode();
+            } else {
+                setUpDesktopMode();
+            }
+            button.editProperties({ isActive: true });
+            playSound(OPEN_SOUND, OPEN_SOUND_VOLUME, MyAvatar.position, true);
+            createMarker();
+            // close tablet so user will be left in state where they can draw
+        }
+    }
+
+    /* ON STOPPING THE SCRIPT: Make sure the marker gets deleted and its variable set back to null 
+    if applicable. Search for any unreferenced markers and delete if found. */
+    function appEnding() {
+        cleanUp();
+        button.clicked.disconnect(onClicked);
+        tablet.removeButton(button);
+        if (controllerMapping) {
+            controllerMapping.disable();
+        }
+    }
+
+    /* CLEANUP: Remove marker, search for any unreferenced markers to clean up */
+    function cleanUp() {
+        if (controllerMapping) {
+            controllerMapping.disable();
+        }
+        if (marker) {
+            Entities.deleteEntity(marker);
+            marker = null;
+        }
+        button.editProperties({ isActive: false });
+        MyAvatar.getAvatarEntitiesVariant().forEach(function(avatarEntity) {
+            var name = Entities.getEntityProperties(avatarEntity.id, 'name').name;
+            if (name === "Draw App Marker") {
+                Entities.deleteEntity(avatarEntity);
+            }
+        });
+    }
+
+    /* WHEN TOGGLING DISPLAY MODE: Set variable to track which method to use to draw lines */
+    function displayModeChange() {
+        if (HMD.active) {
+            closeDesktopMode();
+            setUpHMDMode();
+        } else {
+            closeHMDMode();
+            setUpDesktopMode();
+        }
+    }
+
+    /* SET UP HMD MODE: create controller mapping to listen for trigger presses */
+    function setUpHMDMode() {
+        if (controllerMapping) {
+            controllerMapping.enable();
+        }
+    }
+
+    /* SET UP DESKTOP MODE: Listen for mouse presses */
+    function setUpDesktopMode() {
+        Controller.mousePressEvent.connect(mousePressed);
+        Controller.mouseMoveEvent.connect(mouseContinueLine);
+        Controller.mouseReleaseEvent.connect(mouseReleased);
+    }
+
+    /* CLOSE HMD MODE: Remove controller mapping */
+    function closeHMDMode() {
+        if (controllerMapping) {
+            controllerMapping.disable();
+        }
+    }
+
+    /* CLOSE DESKTOP MODE: Stop listening for mouse presses */
+    function closeDesktopMode() {
+        try {
```

Comment:
instead of using `try catch` here, I recommend setting up a `bool` flag (i usually call it `mouseEventsWired` or similar) that's set to `false` by default, then set to `true` when the events are hooked up.

then you can use that flag to ensure you don't double-connect events, or attempt to double-disconnect events. without doing this, you can run into strange errors that manifest in strange ways.

please shoot me a slack DM if you'd like more details on this :)